src/target/riscv/riscv.c

```diff
@@ -6385,6 +6392,65 @@ bool reg_is_initialized(const struct reg *reg)
 	return true;
 }
 
+static struct target *get_target_from_reg(const struct reg *reg)
+{
+	assert(reg_is_initialized(reg));
+	return ((const riscv_reg_info_t *)reg->arch_info)->target;
+}
+
+static int riscv011_reg_get(struct reg *reg)
+{
+	struct target * const target = get_target_from_reg(reg);
+	RISCV_INFO(r);
+	assert(r);
+	assert(r->dtm_version == DTM_DTMCS_VERSION_0_11);
+	riscv_reg_t value;
+	const int result = r->get_register(target, &value, reg->number);
+	if (result != ERROR_OK)
+		return result;
+	buf_set_u64(reg->value, 0, reg->size, value);
+	return ERROR_OK;
+}
+
+static int riscv011_reg_set(struct reg *reg, uint8_t *buf)
+{
+	const riscv_reg_t value = buf_get_u64(buf, 0, reg->size);
+	struct target * const target = get_target_from_reg(reg);
+	RISCV_INFO(r);
+	assert(r);
+	assert(r->dtm_version == DTM_DTMCS_VERSION_0_11);
+	if (reg->number == GDB_REGNO_TDATA1 || reg->number == GDB_REGNO_TDATA2) {
+		r->manual_hwbp_set = true;
+		/* When enumerating triggers, we clear any triggers with DMODE set,
+		 * assuming they were left over from a previous debug session. So make
+		 * sure that is done before a user might be setting their own triggers.
+		 */
+		if (riscv_enumerate_triggers(target) != ERROR_OK)
+			return ERROR_FAIL;
+	}
+	return r->set_register(target, reg->number, value);
+}
+
+static struct reg_arch_type *gdb_regno_reg_type(const struct target *target,
+		uint32_t regno)
+{
+	RISCV_INFO(info);
+	assert(info);
+	if (info->dtm_version == DTM_DTMCS_VERSION_0_11) {
+		static struct reg_arch_type riscv011_reg_type = {
+			.get = riscv011_reg_get,
+			.set = riscv011_reg_set
+		};
+		return &riscv011_reg_type;
+	}
+
+	static struct reg_arch_type riscv013_reg_type = {
+		.get = riscv013_reg_get,
+		.set = riscv013_reg_set
+	};
+	return &riscv013_reg_type;
+}
+
```

Comment:
Details:

- I would recommend to move the struct definitions outside of the function - for consistency with other targets. It looks all the other targets follow the convention of having `static const struct reg_arch_type *some_name = ...`.

- A safety assertion could be added for `info->dtm_version`, if you like.

```suggestion
static const struct reg_arch_type riscv011_reg_type = {
	.get = riscv011_reg_get,
	.set = riscv011_reg_set
};

static struct reg_arch_type riscv013_reg_type = {
	.get = riscv013_reg_get,
	.set = riscv013_reg_set
};

static const struct reg_arch_type *gdb_regno_reg_type(const struct target *target,
		uint32_t regno)
{
	RISCV_INFO(info);
	assert(info);
	switch (info->dtm_version)
	{
		case DTM_DTMCS_VERSION_0_11:
			return &riscv011_reg_type;
		case DTM_DTMCS_VERSION_1_0:
			return &riscv013_reg_type;
		default:
			assert(false);
			return NULL;
	}
}

```

Reply:
I'd like to address it in #1084 if you don't mind.

Reply:
Sure - thanks.